Pre-authorize child search transport actions

slobodanadamovic opened the pull request:
This PR aims to improve authorization performance of `indices:data/read/search`
action by avoiding authorizing the transport child actions on every node. The focus 
is on index search child actions since they are accessing just a subset of parent's indices. 

Some optimizations already exist which allow the children of authorized parent actions 
_on the same node_ to skip authorization (https://github.com/elastic/elasticsearch/pull/77221).
This PR adds ability to do the same optimization, but for search child actions
that are executed on remote nodes in the _same cluster_.

slobodanadamovic committed:
Resolve cluster alias of connections targeting remote cluster nodes

slobodanadamovic committed:
Test remote cluster alias resolution of proxied connection

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-search-pre-authorization

slobodanadamovic committed:
Pre-authorize index search child actions.

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-search-pre-authorization

slobodanadamovic committed:
Skip child action authorizations in RBACEngine

slobodanadamovic committed:
Revert last change to run.gradle and update RBACEngine

slobodanadamovic committed:
Fix NPE.

slobodanadamovic committed:
Assert no pre-authorization header is set for requests to remote cluster

slobodanadamovic committed:
Fix leakage of parent authorization header and improve debug logs

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-search-pre-authorization

slobodanadamovic commented:
@elasticsearchmachine run elasticsearch-ci/part-3

slobodanadamovic commented:
@elasticsearchmachine run elasticsearch-ci/part-3-fips

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java:
```diff
@@ -262,6 +264,7 @@ public void authorize(
 
         final AuthorizationContext enclosingContext = extractAuthorizationContext(threadContext, action);
+        final ParentActionAuthorization parentAuthorization = extractParentAuthorization(threadContext);
 
```
Parent authz on the remote node is conveyed in a regular header rather than a transient one, so there is no need to store it like this because it is not cleared by the `threadContext.newStoredContext` call below (line 276).

I think this oversight is causing the multi-cluster search failures https://gradle-enterprise.elastic.co/s/p6jiurxwagew6 . The "permanent" header is set and is also carried over remote connections.

albertzaharovits replied:
> Parent authz on the remote node is conveyed in a regular header rather than a transient one, so there is no need to store it like this because it is not cleared by the

We talked it over in a sync meeting.
We concluded that it's preferable from a code structure point of view to access (in this case read) the threadContext in the `AuthorizationService` rather than the `AuthorizationEngine` (`RBACEngine`).

slobodanadamovic committed:
Avoid setting parent authorization header for local node requests

slobodanadamovic committed:
Small log changes.

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-search-pre-authorization

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-search-pre-authorization

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-search-pre-authorization

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java:
```diff
@@ -508,10 +527,17 @@ private void handleIndexActionAuthorizationResult(
-        securityContext.putIndicesAccessControl(result.getIndicesAccessControl());
+        securityContext.putIndicesAccessControl(indicesAccessControl);
+
+        if (result.isGranted() && AuthorizationUtils.isParentAction(action)) {
```
I propose we move the `result.isGranted` in an "assert" (or better, ditch it completely), as it's redundant here (this method is not called in case authz fails).
It's unfortunate that the `IndicesAccessControl` object even has an `isGranted` flag.

Also, more importantly,  `AuthorizationUtils.isParentAction` is very generic looking, in a very generic utils class, but it's actually checking a very specific condition: if this action can have all it's child actions "bypass" their authorization.
My suggestion is something of the sorts: `securityContext.maybeSkipChildreanActionAuthorization`.

slobodanadamovic replied:
Good point! I'll rewrite this along with other changes.

albertzaharovits commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/AuthorizationEngine.java:
```diff
@@ -667,6 +699,133 @@ public IndicesAccessControl getIndicesAccessControl() {
+
+        private static final ParseField VERSION = new ParseField("version");
+        private static final ParseField ACTION = new ParseField("action");
+        private static final ParseField ID = new ParseField("id");
```
I think the `id` is more confusing than useful.
For debug purposes, the node id and the task id are comparatively more useful, I think. Those are already available (eg `request.getParentTask`) from other sources, no need to serialize them in this header.

The version field also has the potential to be confusing because there are already similar versions around, i.e. node version, subject (authn version). If you ask me, I would not add it, but I don't have a strong preference.

slobodanadamovic replied:
I agree, ID is not necessary. I've added it initially to help me with debugging. 

The version is not necessary as well. I was mostly thinking it's useful for potential future changes. But we can easily add it when needed.

Removed both `id` and `version` in https://github.com/elastic/elasticsearch/pull/91886/commits/8428ba2a2dc10c62d6b8dd09ddf489ce2df7e4a0

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java:
```diff
@@ -44,6 +62,33 @@
+            "indices:data/read/search[can_match]",
+            "indices:data/read/search[can_match][n]"
+        )
+    );
```
We talked this over and concluded that only the following child actions would be covered by pre-authorization:
indices:data/read/search[free_context]
indices:data/read/search[phase/dfs]
indices:data/read/search[phase/query]
indices:data/read/search[phase/query/id]
indices:data/read/search[phase/fetch/id]
indices:data/read/search[can_match]
indices:data/read/search[can_match][n]

slobodanadamovic replied:
Adjusted in https://github.com/elastic/elasticsearch/pull/91886/commits/d51306f9275bb915979e606d718c28b3210d04d0

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java:
```diff
@@ -153,4 +198,182 @@ public static void switchUserBasedOnActionOriginAndExecute(
+        } else if (existing.getAction().equals(context.getAction()) == false) {
+            throw new AssertionError("parent authorization context already set in thread context");
+        }
+    }
```
Nit, please create a class for `PreAuthorization` don't dump them in the `AuthorizationUtils` class.

slobodanadamovic replied:
I introduced new [PreAuthorizationUtil](https://github.com/elastic/elasticsearch/blob/bed1362aeac80b4988b3c1f10796a9e0803ce838/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/PreAuthorizationUtil.java) which defines all methods for handling pre-authorizations.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java:
```diff
@@ -153,4 +198,182 @@ public static void switchUserBasedOnActionOriginAndExecute(
+            return;
+        }
+
+        final IndicesAccessControl indicesAccessControl = parentContext.getIndicesAccessControl();
```
You've checked for DLS/FLS in other places, but I think this is the best place do it: i.e. when deciding to pre-authorize.

slobodanadamovic replied:
I've checked on both receiving (remote node) and sending side (coordinating node).
Reason is to make sure DLS/FLS is not enabled on all nodes. If I were to check it only here (on coordinating node) it is possible that roles are configured differently on remote nodes and we would potentially pre-authorize child action without respecting DLS/FLS.

albertzaharovits replied:
I agree, checking that the role doesn't have DLS/FLS on both the parent action (just before setting the header), and on the child action (just before "bypassing" regular authorization sound like the best and cautious option.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java:
```diff
@@ -153,4 +198,182 @@ public static void switchUserBasedOnActionOriginAndExecute(
+        // Checks if the parent context has already successfully authorized access to the child's indices.
+        if (Arrays.stream(indices).allMatch(indicesAccessControl::hasIndexPermissions) == false) {
+            return;
+        }
```
Do you have any details about why these checks are needed? I haven't looked closely because the code will probably change, but I was curious if you have a clear case for them.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java:
```diff
@@ -153,4 +198,182 @@ public static void switchUserBasedOnActionOriginAndExecute(
+        if (indices == null || indices.length == 0 || Arrays.equals(IndicesAndAliasesResolverField.NO_INDICES_OR_ALIASES_ARRAY, indices)) {
+            // No indices to check
+            return;
+        }
```
Do you have any details about why these checks are needed? I haven't looked closely because the code will probably change, but I was curious if you have a clear case for them.

albertzaharovits replied:
Okay, agreed, it makes sense to go with the regular authz if the indices list is empty for any reason.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java:
```diff
@@ -153,4 +198,182 @@ public static void switchUserBasedOnActionOriginAndExecute(
+        // indices access control as well, which we want to avoid with this optimization.
+        if (role.hasFieldOrDocumentLevelSecurity()) {
+            return;
+        }
```
The role DLS permission check is redundant. It's enough to check for DLS on the indices permission.

slobodanadamovic replied:
I completely agree, but I'd prefer to keep it as a sanity check.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java:
```diff
@@ -153,4 +198,182 @@ public static void switchUserBasedOnActionOriginAndExecute(
+        if (connection.getNode().equals(clusterState.nodes().getLocalNode())) {
+            // Child actions targeting local (same) node are handled differently and already pre-authorized by parent action.
+            return;
+        }
```
We discussed in sync that this is an issue, it's a source of confusion that remote child actions are pre-authorized but local ones go the current full-authz flow.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:
```diff
@@ -122,19 +126,40 @@ public <T extends TransportResponse> void sendRequest(
+                                clusterService.state()
+                            );
+                            sendWithUser(connection, action, request, options, handler, sender);
+                        }
```
We talked this over in a sync meeting:

My understanding is that the **persistent** pre-authz header will be set and implicitly copied to all the nodes in the same cluster (no remote cluster). There will be no additional **transient** pre-authz header.

albertzaharovits reviewed (changes requested):
After discussions I think the plan is to set the **persistent** header once, after authorizing a search action, and remove it for connections to remote clusters. The header will be interpreted during the authz of some (most) search child actions, in which case authz will pass faster (without inspecting the indices list of search child actions).

**In addition**, I would also like to propose that we clear the **persistent** header after the first child action (i.e. the pre-authorization only works for the immediate children).
I'm worried that some child actions, for example the actions that rewrite a search query (for e.g. for the terms lookup query type), might have themselves child actions that can be erroneously interpreted as pre-authorized.
We simply don't have a good way to track child of children actions.
I can be persuaded to defer this but I would very much prefer we implement this in the first iteration.

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-search-pre-authorization

slobodanadamovic committed:
Remove scroll actions from pre-authorization.

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-search-pre-authorization

slobodanadamovic committed:
Remove id and version properties from parent authorization

slobodanadamovic committed:
Rework pre-authorization:

- to set parent authorization only once as persistent header
- and remove it from context when it's not needed on the outbound

slobodanadamovic committed:
Reuse newly added method.

slobodanadamovic committed:
Apply spotless.

slobodanadamovic commented:
@elasticsearchmachine run elasticsearch-ci/bwc

slobodanadamovic commented:
@elasticmachine run elasticsearch-ci/bwc

slobodanadamovic committed:
Minor cosmetic changes.

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-search-pre-authorization

slobodanadamovic committed:
Code cleanup, javadoc, renaming...

slobodanadamovic committed:
Move all pre-authorization related code to PreAuthorizationUtil

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-search-pre-authorization

slobodanadamovic committed:
Tests for AuthorizationService and RBACEngine

slobodanadamovic committed:
Use constants instead of string literals.

slobodanadamovic committed:
Test SecurityContext

slobodanadamovic committed:
New test case when role has DLS/FLS configured

slobodanadamovic committed:
Small changes to simplify testing of SecurityServerTransportInterceptor

slobodanadamovic committed:
Use resolver consistently in security interceptor

albertzaharovits commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/SecurityContext.java:
```diff
@@ -186,6 +204,24 @@ public void executeAfterRewritingAuthentication(Consumer<StoredContext> consumer
+                    && threadContext.getHeader(k) == null) {
+                    threadContext.putHeader(k, v);
+                }
+            });
```
This does more than clearing the parent authz header. It also resets transient headers.

If I'm right, I'm not sure how the code in `SecurityServerTransportInterceptor#sendRequestInner` can still work correctly, because it inspects transient values that this method clears.

slobodanadamovic replied:
Changed remove method in the commit https://github.com/elastic/elasticsearch/pull/91886/commits/db6cc06232db00196cef1ae6f1162340dd0f870f and adjusted `ThreadContext#getTransientHeaders` in https://github.com/elastic/elasticsearch/pull/91886/commits/091612556038a80c443b3cceb83d6c8ddacd9683.

albertzaharovits replied:
It still clears the `ThreadContextStruct#isSystemContext` and the `warningHeadersSize`.
There is a method `ThreadContext#newStoredContext` that clears specified transient headers.
Maybe you could adapt that to take another parameter for the persistent headers to clear, and then reuse that method here in `SecurityContext`?

slobodanadamovic replied:
Sorry for this slip again and thanks for the suggestion. I definitely need to make a full copy of the `ThreadContextStruct` and `ThreadContext#newStoredContext` looks fitting since is only ever used in `AuthorizationService`.

slobodanadamovic replied:
I've adapted `ThreadContext#newStoredContext` and added additional collection for request headers.
I also introduced new `ThreadContext#newStoredContextPreservingResponseHeaders` method which does the same (clears given transient and request headers) but also preserves response headers.

https://github.com/elastic/elasticsearch/pull/91886/commits/455e0dfcc125dbf6ced20142dba6fa1edede52b0

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/PreAuthorizationUtil.java:
```diff
@@ -0,0 +1,209 @@
+            return;
+        }
+
+        final Optional<ParentActionAuthorization> existingParentAuthorization = Optional.ofNullable(
```
I don't think this usage of `Optional` improves the comprehension of the code or brings any desirable enhancement. I suggest it be removed.

slobodanadamovic replied:
Agreed. I will remove it.

slobodanadamovic replied:
https://github.com/elastic/elasticsearch/pull/91886/commits/e5bc23043cf55596ce6bcd9383c32766f70cfcbf

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/PreAuthorizationUtil.java:
```diff
@@ -0,0 +1,209 @@
+        // indices access control as well, which we want to avoid with this optimization.
+        if (role.hasFieldOrDocumentLevelSecurity()) {
+            return;
+        }
```
I think this is too general.
The role can have DLS/FLS permissions defined but which don't cover the indices in the request.
Something like `indicesAccessControl.getFieldAndDocumentLevelSecurityUsage().hasFieldOrDocumentLevelSecurity()` would be more precise.

slobodanadamovic replied:
That makes sense, but going with the approach to check if `indicesAccessControl` has DLS or FLS would mean that we have to check access for all indices. Since we plan to make IAC computed lazily, it would mean that we would always access IAC on a coordinating node and we want to avoid that.

albertzaharovits replied:
Yeah, that's a good point. Okay lets keep it as is for now.

slobodanadamovic committed:
Do not remove transient headers when removing parent authorization

slobodanadamovic committed:
Rename PreAuthorizationUtils and write some tests for it

slobodanadamovic committed:
Don't use optional.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/PreAuthorizationUtil.java:
```diff
@@ -0,0 +1,209 @@
+                // Single request can fan-out a child action to multiple nodes in the cluster, e.g. node1 and node2.
+                // Sending a child action to node1 would have already put parent authorization in the thread context.
+                // To avoid attempting to pre-authorize the same parent action twice we simply return here
+                // since parent authorization is already set in the thread context.
```
I don't understand this explanation.
Under what circumstances is it expected to execute a parent action as a child of another parent action?

slobodanadamovic replied:
Old comment which does not hold true anymore. I will change it.

slobodanadamovic replied:
Note: The check is still needed due to the double authorization executed by `SecurityActionFilter`.
Here is a long stacktrace on a `fulfilling-cluster` which has a single node:

```
[2022-12-14T16:57:25,314][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [fulfilling-cluster-0] fatal error in thread [elasticsearch[fulfilling-cluster-0][system_critical_read][T#3]], exiting java.lang.AssertionError: authorization for parent action [indices:data/read/search] is already set in the thread context
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.PreAuthorizationUtils.maybeSkipChildrenActionAuthorization(PreAuthorizationUtils.java:109)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.handleIndexActionAuthorizationResult(AuthorizationService.java:526)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.lambda$authorizeAction$11(AuthorizationService.java:484)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$AuthorizationResultListener.onResponse(AuthorizationService.java:988)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$AuthorizationResultListener.onResponse(AuthorizationService.java:954)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:31)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.RBACEngine.authorizeIndexAction(RBACEngine.java:359)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.authorizeAction(AuthorizationService.java:477)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.maybeAuthorizeRunAs(AuthorizationService.java:414)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.lambda$authorize$3(AuthorizationService.java:315)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:31)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.RBACEngine.lambda$resolveAuthorizationInfo$0(RBACEngine.java:148)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.lambda$getRoles$1(CompositeRolesStore.java:201)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.xcore@8.7.0-SNAPSHOT/org.elasticsearch.xpack.core.security.authz.store.RoleReferenceIntersection.lambda$buildRole$0(RoleReferenceIntersection.java:49)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.GroupedActionListener.onResponse(GroupedActionListener.java:55)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.buildRoleFromRoleReference(CompositeRolesStore.java:299)
	at org.elasticsearch.xcore@8.7.0-SNAPSHOT/org.elasticsearch.xpack.core.security.authz.store.RoleReferenceIntersection.lambda$buildRole$1(RoleReferenceIntersection.java:53)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at org.elasticsearch.xcore@8.7.0-SNAPSHOT/org.elasticsearch.xpack.core.security.authz.store.RoleReferenceIntersection.buildRole(RoleReferenceIntersection.java:53)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.getRole(CompositeRolesStore.java:216)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.getRoles(CompositeRolesStore.java:191)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.RBACEngine.resolveAuthorizationInfo(RBACEngine.java:145)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.authorize(AuthorizationService.java:317)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.lambda$applyInternal$4(SecurityActionFilter.java:159)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$MappedActionListener.onResponse(ActionListener.java:127)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authc.AuthenticatorChain.authenticateAsync(AuthenticatorChain.java:93)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authc.AuthenticationService.authenticate(AuthenticationService.java:171)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.applyInternal(SecurityActionFilter.java:155)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.apply(SecurityActionFilter.java:114)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:84)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:61)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.HandledTransportAction$TransportHandler.messageReceived(HandledTransportAction.java:71)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.HandledTransportAction$TransportHandler.messageReceived(HandledTransportAction.java:67)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler$1.doRun(SecurityServerTransportInterceptor.java:319)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler$2.onResponse(SecurityServerTransportInterceptor.java:366)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler$2.onResponse(SecurityServerTransportInterceptor.java:363)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$1.onResponse(AuthorizationService.java:605)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$1.onResponse(AuthorizationService.java:599)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.interceptor.DlsFlsLicenseRequestInterceptor.intercept(DlsFlsLicenseRequestInterceptor.java:100)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$1.onResponse(AuthorizationService.java:603)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$1.onResponse(AuthorizationService.java:599)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.interceptor.FieldAndDocumentLevelSecurityRequestInterceptor.intercept(FieldAndDocumentLevelSecurityRequestInterceptor.java:81)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.interceptor.SearchRequestInterceptor.intercept(SearchRequestInterceptor.java:25)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$1.onResponse(AuthorizationService.java:603)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$1.onResponse(AuthorizationService.java:599)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.interceptor.BulkShardRequestInterceptor.intercept(BulkShardRequestInterceptor.java:85)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$1.onResponse(AuthorizationService.java:603)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$1.onResponse(AuthorizationService.java:599)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.interceptor.FieldAndDocumentLevelSecurityRequestInterceptor.intercept(FieldAndDocumentLevelSecurityRequestInterceptor.java:81)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.interceptor.ShardSearchRequestInterceptor.intercept(ShardSearchRequestInterceptor.java:24)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$1.onResponse(AuthorizationService.java:603)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$1.onResponse(AuthorizationService.java:599)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.interceptor.IndicesAliasesRequestInterceptor.intercept(IndicesAliasesRequestInterceptor.java:122)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$1.onResponse(AuthorizationService.java:603)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$1.onResponse(AuthorizationService.java:599)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.interceptor.FieldAndDocumentLevelSecurityRequestInterceptor.intercept(FieldAndDocumentLevelSecurityRequestInterceptor.java:81)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.interceptor.UpdateRequestInterceptor.intercept(UpdateRequestInterceptor.java:27)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$1.onResponse(AuthorizationService.java:603)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$1.onResponse(AuthorizationService.java:599)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.interceptor.ResizeRequestInterceptor.intercept(ResizeRequestInterceptor.java:93)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.runRequestInterceptors(AuthorizationService.java:599)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.handleIndexActionAuthorizationResult(AuthorizationService.java:584)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.lambda$authorizeAction$11(AuthorizationService.java:484)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$AuthorizationResultListener.onResponse(AuthorizationService.java:988)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$AuthorizationResultListener.onResponse(AuthorizationService.java:954)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:31)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.RBACEngine.lambda$authorizeIndexAction$3(RBACEngine.java:385)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ListenableFuture.notifyListenerDirectly(ListenableFuture.java:113)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ListenableFuture.addListener(ListenableFuture.java:55)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ListenableFuture.addListener(ListenableFuture.java:41)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService$CachingAsyncSupplier.getAsync(AuthorizationService.java:1034)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.RBACEngine.authorizeIndexAction(RBACEngine.java:365)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.authorizeAction(AuthorizationService.java:477)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.maybeAuthorizeRunAs(AuthorizationService.java:414)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.lambda$authorize$3(AuthorizationService.java:315)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:31)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.RBACEngine.lambda$resolveAuthorizationInfo$0(RBACEngine.java:148)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.lambda$getRoles$1(CompositeRolesStore.java:201)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.xcore@8.7.0-SNAPSHOT/org.elasticsearch.xpack.core.security.authz.store.RoleReferenceIntersection.lambda$buildRole$0(RoleReferenceIntersection.java:49)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.GroupedActionListener.onResponse(GroupedActionListener.java:55)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.lambda$buildThenMaybeCacheRole$7(CompositeRolesStore.java:369)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.buildRoleFromDescriptors(CompositeRolesStore.java:514)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.buildThenMaybeCacheRole(CompositeRolesStore.java:350)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.lambda$buildRoleFromRoleReference$4(CompositeRolesStore.java:288)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.RoleDescriptorStore.lambda$resolveRoleNames$3(RoleDescriptorStore.java:156)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.RoleDescriptorStore.lambda$loadRoleDescriptorsAsync$8(RoleDescriptorStore.java:218)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:31)
	at org.elasticsearch.xcore@8.7.0-SNAPSHOT/org.elasticsearch.xpack.core.common.IteratingActionListener.onResponse(IteratingActionListener.java:141)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.RoleDescriptorStore.lambda$loadRoleDescriptorsAsync$12(RoleDescriptorStore.java:245)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:162)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.NativeRolesStore$4.onResponse(NativeRolesStore.java:376)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.store.NativeRolesStore$4.onResponse(NativeRolesStore.java:372)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:31)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.client.internal.node.NodeClient$SafelyWrappedActionListener.onResponse(NodeClient.java:160)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.tasks.TaskManager$1.onResponse(TaskManager.java:208)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.tasks.TaskManager$1.onResponse(TaskManager.java:202)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:31)
	at org.elasticsearch.security@8.7.0-SNAPSHOT/org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.lambda$applyInternal$2(SecurityActionFilter.java:165)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListener$DelegatingFailureActionListener.onResponse(ActionListener.java:245)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionListenerResponseHandler.handleResponse(ActionListenerResponseHandler.java:43)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleResponse(TransportService.java:1362)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TransportService$DirectResponseChannel.processResponse(TransportService.java:1461)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1432)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:41)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.ChannelActionListener.onResponse(ChannelActionListener.java:39)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.support.ChannelActionListener.onResponse(ChannelActionListener.java:20)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:50)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:47)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$3.doRun(ActionRunnable.java:72)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:924)
	at org.elasticsearch.server@8.7.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1589)
`

albertzaharovits replied:
Thank you for the stacktraces.
However, honestly, I don't fully understand what's going on here, and I think it's worth investigating.
Can you please make this change and let the CI run with it:
```diff --git a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/PreAuthorizationUtils.java b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/PreAuthorizationUtils.java
index 6f2d9f62c09..aec63d31584 100644
--- a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/PreAuthorizationUtils.java
+++ b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/PreAuthorizationUtils.java
@@ -97,22 +97,13 @@ public final class PreAuthorizationUtils {

         final ParentActionAuthorization existingParentAuthorization = securityContext.getParentAuthorization();
         if (existingParentAuthorization != null) {
-            if (existingParentAuthorization.action().equals(parentAction)) {
-                // This can happen if authorization is executed more than once for the same parent action.
-                // At the moment of writing, SecurityActionFilter can be executed twice for the same action
-                // when the action receiving and the action executing node is the same "local node".
-                if (logger.isDebugEnabled()) {
-                    logger.debug("authorization for parent action [" + parentAction + "] is already set in the thread context");
-                }
-            } else {
-                throw new AssertionError(
-                    "found parent authorization for action ["
-                        + existingParentAuthorization.action()
-                        + "] while attempting to set authorization for new parent action ["
-                        + parentAction
-                        + "]"
-                );
-            }
+            throw new AssertionError(
+                "found parent authorization for action ["
+                    + existingParentAuthorization.action()
+                    + "] while attempting to set authorization for new parent action ["
+                    + parentAction
+                    + "]"
+            );
         } else {
             if (logger.isDebugEnabled()) {
                 logger.debug("adding authorization for parent action [" + parentAction + "] to the thread context");
```
I don't think double-authz as I know it happens for search actions (only for `TransportMasterNodeAction`), and the stacktraces look from a remote cluster search which shouldn't see the parent authz header in recent iterations of this PR.

slobodanadamovic replied:
> the stacktraces look from a remote cluster search which shouldn't see the parent authz header in recent iterations of this PR.

Correct. Running remote cluster search never passes parent authz from querying cluster. The header is added on the fulfilling side. There are two logged messages on the fulfilling node:  `adding authorization for parent action...` and `authorization for parent action is already set...`

I've pushed the changes: https://github.com/elastic/elasticsearch/pull/91886/commits/07d6ad8331eed6cbe5bf2f204190e717745f6690

Reproduce with:
```
./gradlew ':x-pack:qa:multi-cluster-search-security:legacy-with-basic-license:querying-cluster' --tests "org.elasticsearch.xpack.security.MultiClusterSearchWithSecurityYamlTestSuiteIT.test" -Dtests.seed=B4A0F28E6AA1D471 -Dtests.locale=sl-SI -Dtests.timezone=America/Rainy_River -Druntime.java=17
```

slobodanadamovic commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/PreAuthorizationUtil.java:
```diff
@@ -0,0 +1,209 @@
+        if (parentRequest instanceof IndicesRequest == false) {
+            // We can only pre-authorize indices request.
+            return;
+        }
```
Agreed.

slobodanadamovic replied:
https://github.com/elastic/elasticsearch/pull/91886/commits/407b3e3555eb99ccb8ed73a394e4f46eae087e3d

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/PreAuthorizationUtil.java:
```diff
@@ -0,0 +1,209 @@
+        if (parentAuthorization == null) {
+            // Nothing to remove.
+            return false;
+        }
```
nit: better make the remove a noop if the header is not present, and remove this "nothing to remove" check.

slobodanadamovic replied:
I'm not sure if I can get rid of this `null` check since I access it at line 155:
https://github.com/elastic/elasticsearch/blob/72f0951379d35ea66c314ad57c0da75bd0f43b62/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/PreAuthorizationUtil.java#L155

but I like the idea to make the remove method noop if the header is not set.

albertzaharovits reviewed (commented):
I'll do another pass after you go over the comments.

slobodanadamovic committed:
Remove unnecessary instance check

slobodanadamovic committed:
Nit: Rename method

slobodanadamovic commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/PreAuthorizationUtil.java:
```diff
@@ -0,0 +1,209 @@
+        }
+    }
+
+    private static boolean isChildActionWhitelistedForParent(final String parent, final String child) {
```
https://github.com/elastic/elasticsearch/pull/91886/commits/dd234a3a04d559447deb249dade7a7840c92f678

slobodanadamovic committed:
Return unmodifiable copy of transient headers
to avoid changes which would not affect thread local

slobodanadamovic committed:
Update comment.

slobodanadamovic committed:
More tests.

slobodanadamovic commented:
@albertzaharovits I've addressed some of the comments and for others I replied. Would appreciate if you could take another look.

albertzaharovits commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/AuthorizationEngine.java:
```diff
@@ -667,6 +697,103 @@ public IndicesAccessControl getIndicesAccessControl() {
+        );
+        static {
+            PARSER.declareString(constructorArg(), ACTION);
+        }
```
nit: It looks like the `ParentActionAuthorization` record is never passed around in "XContent" form (such as request or response bodies).
I would avoid adding code that's only exercised in tests, so I would suggest you remove the parser.

slobodanadamovic replied:
Agreed. I've removed the XContent logic. https://github.com/elastic/elasticsearch/pull/91886/commits/7b5e28f9c94535c4355ce297acae73a526304ff4

albertzaharovits commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/AuthorizationEngine.java:
```diff
@@ -667,6 +697,103 @@ public IndicesAccessControl getIndicesAccessControl() {
+     */
+    record ParentActionAuthorization(String action) implements Writeable, ToXContent {
+
+        public static final String THREAD_CONTEXT_KEY = "_xpack_security_authorization";
```
nit: rename to "_xpack_security_parent_action_authz"

slobodanadamovic replied:
Good suggestion. I've renamed it https://github.com/elastic/elasticsearch/pull/91886/commits/95e1054a394274190a1f12dc683a62e779dc8b5f

albertzaharovits commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/AuthorizationEngine.java:
```diff
@@ -667,6 +697,103 @@ public IndicesAccessControl getIndicesAccessControl() {
+            context.putHeader(THREAD_CONTEXT_KEY, header);
+        }
+
+        private void ensureContextDoesNotContainAuthorization(ThreadContext context) {
```
nit: it looks like the `ThreadContext#putHeader` already has checks to not overwrite headers, so this check looks redundant.

slobodanadamovic replied:
Good point. I've removed the check.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java:
```diff
@@ -556,11 +564,7 @@ private void handleIndexActionAuthorizationResult(
-                authzInfo,
-                result.getIndicesAccessControl()
-            );
+            final AuthorizationContext authzContext = new AuthorizationContext(action, authzInfo, indicesAccessControl);
```
This looks redundant, you can reuse the "parentContext" above (which is not really a "parent' in the context it has been named like that, it's a parent inside "marybeSkipChildrenActionAuthorization".

slobodanadamovic replied:
It is. I've renamed `parentContext` to `authzContext` and removed this line as well. https://github.com/elastic/elasticsearch/pull/91886/commits/c3cc31a635753fcaf6b374b53e3dc46a0f27bdaf

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:
```diff
@@ -155,6 +211,12 @@ private <T extends TransportResponse> void sendWithUser(
+            if (remoteClusterAliasResolver.apply(connection).isPresent()) {
+                assert false : "parent authorization header should not be set for remote cluster requests";
+            }
+        }
```
nit: I find this distracting, please remove or replace with:
```
assert securityContext.getParentAuthorization() == null || remoteClusterAliasResolver.apply(connection).isPresent() == false : "parent authorization header should not be set for remote cluster requests"
```

slobodanadamovic replied:
Replaced https://github.com/elastic/elasticsearch/pull/91886/commits/91f5f4a68b1510c007d32d488d24c5e6634ca620

albertzaharovits commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/SecurityContext.java:
```diff
@@ -86,6 +86,24 @@ public AuthorizationEngine.AuthorizationInfo getAuthorizationInfoFromContext() {
+        try {
+            parentAuthorization.writeToThreadContext(threadContext);
+        } catch (IOException e) {
+            throw new AssertionError("failed to write parent authorization to the thread context", e);
```
nit: maybe stick with the `UnckedIOException(e)` pattern, unless I'm missing the reasoning why this should be an assertion rather than a hard exception.

slobodanadamovic committed:
Fix parent authorization header from ThreadContext
by restoring original and preserving response headers

slobodanadamovic committed:
Make test constructor package protected

slobodanadamovic committed:
Change if to assert only

slobodanadamovic committed:
Break line which was causing style check to fail.

slobodanadamovic committed:
Rename parent auhtorization header.

slobodanadamovic committed:
Address review comments:

- remove ToXContent implementation
- remove redundant assertion

slobodanadamovic committed:
Rename parentContext variable to authzContext and reuse on other places.

albertzaharovits commented on server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java:
```diff
@@ -510,6 +544,13 @@ public <T> T getTransient(String key) {
+     */
+    public Map<String, Object> getTransientHeaders() {
+        return Collections.unmodifiableMap(threadLocal.get().transientHeaders);
+    }
```
nit: looks like you don't need this anymore.

slobodanadamovic replied:
I kind of do, as it helps me with testing the context stashing. I can make it package protected?

slobodanadamovic committed:
Temporary changes for debugging double authz

slobodanadamovic committed:
Remove leniency when authorizing parent action
and clear parent authorization header so it does not get leaked

slobodanadamovic commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/PreAuthorizationUtils.java:
```diff
@@ -0,0 +1,203 @@
+            return false;
+        }
+
+        Role role = RBACEngine.maybeGetRBACEngineRole(childAuthorizationInfo);
```
Good point. I've added a `null` check.

slobodanadamovic committed:
Fix failing tests after last change to not "leak" authz header.

slobodanadamovic commented:
@elasticmachine run elasticsearch-ci/bwc

slobodanadamovic commented:
@elasticmachine update branch please

elasticmachine committed:
Merge branch 'main' into sa-search-pre-authorization

albertzaharovits reviewed (approved):
LGTM

slobodanadamovic committed:
Update docs/changelog/91886.yaml

elasticsearchmachine commented:
Hi @slobodanadamovic, I've created a changelog YAML for you.
